### PR TITLE
Keep subclassing in apply

### LINKED ISSUE #19822: Subclassing is lost when using DataFrame.apply
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

# Define a subclass of Series
class ExtendedSeries( pd.Series ):
    @property
    def _constructor(self):
        return ExtendedSeries

# Define a subclass of DataFrame that slices into the ExtendedSeries class
class ExtendedFrame( pd.DataFrame ):
    @property
    def _constructor(self):
        return ExtendedFrame
    _constructor_sliced = ExtendedSeries

# Declare instance of ExtendedFrame
df = ExtendedFrame({"c1": [1, 2, 3, 4], "c2": [2, 3, 4, 5], "c3": [3, 4, 5, 6]})

print(type(df))
# <class '__main__.ExtendedFrame'>

print(type(df.iloc[0]))
# <class '__main__.ExtendedSeries'>

for x in df.apply(lambda x: type(x), axis=1):
    print(x)
#<class 'pandas.core.series.Series'>
#<class 'pandas.core.series.Series'>
#<class 'pandas.core.series.Series'>
#<class 'pandas.core.series.Series'>
```

#### Problem description

The class type of the _applied_ objects should still be `<class '__main__.ExtendedSeries'>` instead of going back to `<class 'pandas.core.series.Series'>`.

The issue here is that in `pandas.core.apply`, whenever the returning objects are generated, they are declared as `DataFrame` and `Series` instead of `self.obj._constructor` and `self.obj._constructor_slice`.

This behaviour makes it difficult to properly work with subclasses from pandas. 

#### Expected Output

Following the initial code example, the expected output should be:
```python
import pandas as pd

# Define a subclass of Series
class ExtendedSeries( pd.Series ):
    @property
    def _constructor(self):
        return ExtendedSeries

# Define a subclass of DataFrame that slices into the ExtendedSeries class
class ExtendedFrame( pd.DataFrame ):
    @property
    def _constructor(self):
        return ExtendedFrame
    _constructor_sliced = ExtendedSeries

# Declare instance of ExtendedFrame
df = ExtendedFrame({"c1": [1, 2, 3, 4], "c2": [2, 3, 4, 5], "c3": [3, 4, 5, 6]})

print(type(df))
# <class '__main__.ExtendedFrame'>

print(type(df.iloc[0]))
# <class '__main__.ExtendedSeries'>

for x in df.apply(lambda x: type(x), axis=1):
    print(x)
#<class '__main__.ExtendedSeries'>
#<class '__main__.ExtendedSeries'>
#<class '__main__.ExtendedSeries'>
#<class '__main__.ExtendedSeries'>
```
(note the change in the last printed lines)

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Darwin
OS-release: 17.4.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.21.1
pytest: None
pip: 9.0.1
setuptools: 38.4.0
Cython: None
numpy: 1.14.0
scipy: 0.18.1
pyarrow: None
xarray: None
IPython: 5.4.1
sphinx: 1.6.6
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.1.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.5.1
html5lib: 0.999999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jaumebonet**
When generating new objects on apply, it calls `self.obj._constructor` and `self.obj._constructior_sliced` instead of:  `DataFrame` and `Series`; keeping subclassing.

- [x] closes #19822
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] keeps subclass inside `DataFrame.apply`

### CODE REVIEW

**Comment by jaumebonet** [comment]
Hi @chris-b1,
Thanks!
Yes, I'm trying to pinpoint what exactly has changed... Clearly it has to do with the fact that now it returns `SparseSeries` instead of `Series`, but I'm not familiar with `SparseSeries`, so I'm not sure what difference does that make.

**Comment by jreback** [comment]
can you add some tests in pandas/tests/frame/test_sub_class (and for Series as well)

**Comment by jreback** [comment]
also pls add a release note, other enhancements section is ok

**Comment by jaumebonet** [comment]
Hi @jreback,
I've added a test to check if subclass of `Series` are kept when using apply on `DataFrame`.
`Series` have they own apply function, so the changes do not affect how `Series.apply` behave. 

As per your suggestion, I also added a bullet point in the _other enhancement_ section.

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -295,6 +295,7 @@ Other Enhancements
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
 - Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)
+- :func:``DataFrame.apply`` keeps provides the specified ``Series`` subclass when `DataFrame._constructor_sliced`` is defined (:issue:`19822`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: don't use a private property here, just say when ``Series`` and ``DataFrame``` are subclassed

File: pandas/tests/frame/test_subclass.py [hunk 2]
```diff
@@ -514,3 +514,19 @@ def test_subclassed_wide_to_long(self):
...
+            assert isinstance(row, tm.SubclassedSeries)
+
+        df = tm.SubclassedDataFrame([
+            ['John', 'Doe', 'height', 5.5],
+            ['Mary', 'Bo', 'height', 6.0],
+            ['John', 'Doe', 'weight', 130],
+            ['Mary', 'Bo', 'weight', 150]],
+            columns=['first', 'last', 'variable', 'value'])
+
+        df.apply(lambda x: check_row_subclass(x))
```
Comment (jreback) [inline pandas/tests/frame/test_subclass.py]: you need to check the result of this using ``tm.assert_frame_equal``, just use a simple example of an applied function, testing both returning a list, e.g. ``lambda x: [1, 2, 3]`` and a sub-classed series

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_subclass.py [hunk 3]
```diff
@@ -514,3 +514,55 @@ def test_subclassed_wide_to_long(self):
...
+            ['Mary', 'Bo', 'weight', 150]],
+            columns=['first', 'last', 'variable', 'value'])
+
+        expected2 = tm.SubclassedDataFrame([
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3]])
+
+        expected3 = DesignSeries([[1, 2, 3], [1, 2, 3], [1, 2, 3], [1, 2, 3]])
```
Comment (TomAugspurger) [inline pandas/tests/frame/test_subclass.py]: Looks like this needs to be updated.

File: pandas/tests/frame/test_subclass.py [hunk 4]
```diff
@@ -514,3 +514,59 @@ def test_subclassed_wide_to_long(self):
...
+        expected2 = tm.SubclassedDataFrame([
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3]])
+
+        expected3 = tm.SubclassedSeries([
+            [1, 2, 3],
+            [1, 2, 3],
+            [1, 2, 3],
```
Comment (jreback) [inline pandas/tests/frame/test_subclass.py]: rather than numbering these, can you just do one after the other and use 
```
result = 
expected = 
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -295,8 +295,10 @@ Other Enhancements
 - ``IntervalIndex.astype`` now supports conversions between subtypes when passed an ``IntervalDtype`` (:issue:`19197`)
 - :class:`IntervalIndex` and its associated constructor methods (``from_arrays``, ``from_breaks``, ``from_tuples``) have gained a ``dtype`` parameter (:issue:`19262`)
 - Added :func:`SeriesGroupBy.is_monotonic_increasing` and :func:`SeriesGroupBy.is_monotonic_decreasing` (:issue:`17015`)
+- :func:``DataFrame.apply`` keeps the specified ``Series`` subclass when ``Series`` and ``DataFrame`` subclasses are defined (:issue:`19822`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.23.0.txt]: I find this not so clear. It is about the Series type that is passed to the function?

Comment (jaumebonet) [reply doc/source/whatsnew/v0.23.0.txt]: Yes.  
When `apply`ing a function to a subclassed `DataFrame` I realised I could not call the specific methods of the subclassed `Series` that this `DataFrame` was supossed to generate  through `_constructor_sliced`. As far as I understood, the subclassing of the `Series` was lost during `apply`.  
With these changes the subclass is now kept inside `apply`, so method of the expected `Series` subclass can be used.

Comment (jorisvandenbossche) [reply doc/source/whatsnew/v0.23.0.txt]: OK, can you try to clarify the text a bit? Eg something like "For subclassed DataFrames, DataFrame.apply will now preserve the Series subclass (if defined) when passing the data to the applied function" (but adapt as you like)

Comment (jaumebonet) [reply doc/source/whatsnew/v0.23.0.txt]: sure! I like it as this! :)
will change it!

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**